### PR TITLE
Fix wrong behavior on Android web

### PR DESCRIPTION
**Pull request by staszekscp**
### Details

Fix for[ wrong freeze behavior ](https://github.com/Expensify/App/pull/16581#issuecomment-1516470982)on android web - it introduces wrapper that provides Freeze component to mobile apps (where it is really needed)

Part of [navigation rework](https://github.com/Expensify/App/issues/11768) by Software Mansion

> BUG: Web: ANDROID: LHN is reloading when we open any page from the FAB menu.

The problem was caused by the `Freeze` component - on Android web the the freezing was triggered too early (before screen's transition end). Instead of implementing some hacky solutions I decided to remove the Freeze component from the web flow, because it was causing more problems without giving much perfromance advantage there.

### Fixed Issues

Related to https://github.com/Expensify/App/pull/16581#issuecomment-1516470982

### Tests

- [x] Verify that no errors appear in the JS console

1. Open the app on the mobile web.
1. Open the Fab menu and click on any menu.
1. Observe that images/page are reloaded in the background.

### QA Steps

- [x] Verify that no errors appear in the JS console

1. Open the app on the mobile web.
1. Open the Fab menu and click on any menu.
1. Observe that images/page are reloaded in the background.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>
Sorry for the quality

https://user-images.githubusercontent.com/75020057/234681247-1c0302e6-7773-4874-9e97-f3978238acc1.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by neil-marcellini** [comment]
@staszekscp please follow our [contributing guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md). Were you hired for an issue with an accepted proposal? If so please include those in the description and re-open the PR.

**Comment by staszekscp** [comment]
cc:  Hey @mountiny, could you have a look at that?

**Comment by neil-marcellini** [comment]
Thanks for linking the related issue and now I see that you're from Software Mansion so I'll reopen this. I was going quickly and thought this was a random PR opened unrelated to an issue.

**Review by mountiny — COMMENTED** [review]
@staszekscp You got some lint issues here again, hopefully we will get rid of this with Prettier and typestript haha but make sure to check if the `npm run lint`  is fine before submitting ready for a review

**Comment by mountiny** [comment]
@s77rt could you give this one a look sooner than later, thank you very much 🙇

**Review by neil-marcellini — CHANGES_REQUESTED** [review]

File: src/pages/home/sidebar/SidebarLinks.js
Comment (staszekscp) [reply src/pages/home/sidebar/SidebarLinks.js:194]: Sure, will do!

**Comment by staszekscp** [comment]
Sorry @mountiny for the lint issues, I keep forgetting that, because in previous projects I couldn't commit anything without passing a lint check via husky script :P I'll keep that in mind, though, so hopefully I will pass the lint check next time!

File: src/pages/home/sidebar/SidebarLinks.js
Comment (s77rt) [reply src/pages/home/sidebar/SidebarLinks.js:194]: I will go thought the files in depth but just a quick response. I assume LHNOptionList is exactly the same and the only difference is Freeze is that right? Then It would be better to isolate that component instead as @neil-marcellini suggested. **BUT** why not remove the Freeze component, is it really doing something significantly good? It's anti react's life-cycle and we will remove it on navigation reboot anyway.

Comment (s77rt) [reply src/pages/home/sidebar/SidebarLinks.js]: Just realised this is a part of navigation reboot :grin: I think It was planned to remove Freeze at this point, why we are still using it. Am I missing something here? cc @neil-marcellini @mountiny (and @roryabraham I guess you know about this)

**Comment by neil-marcellini** [comment]
Thanks for the SidebarFreeze update! I think as a general rule we should always keep platform specific components as small as possible.

Please include testing videos as described in the PR author checklist. I'm not able to observe the LHN reloading under the global create menu on main. @parasharrajat is that easy to observe or do we need to add logging or otherwise debug to see if the component is updating or remounting? Also could you please write a couple sentences about the problem and how this fixes it, it's not super clear to me.

I'll let @s77rt take the lead on the reviews and come back after he approves.

**Comment by staszekscp** [comment]
@neil-marcellini It is a bug specific to the navigation rework, that's why it is not possible to notice it on the `main`  :)

**Comment by s77rt** [comment]
@neil-marcellini I still have concerns about the use of Freeze, please check comment https://github.com/Expensify/App/pull/17962#discussion_r1178287355 in case you missed it

File: src/components/FlatList/index.android.js
Comment (s77rt) [inline src/components/FlatList/index.android.js:1]: All of this is a workaround. The bug is upstream and it has been reported already a while ago https://github.com/software-mansion/react-freeze/issues/31

**Review by s77rt — CHANGES_REQUESTED** [review]
- The custom FlatList is a workaround. bug is upstream: https://github.com/software-mansion/react-freeze/issues/31
- Why not just remove Freeze?

cc @neil-marcellini

File: src/components/LHNOptionsList/index.js
Comment (mountiny) [reply src/components/LHNOptionsList/index.js:1]: This is outdated, there was a web file before

**Review by mountiny — COMMENTED** [review]
@s77rt I think we want to keep Freeze to limit the number of rerender in the stack. All the navigation will be stack based so if we wont have Freeze around the LHN, it would be rerendering in the background

**Comment by mountiny** [comment]
@s77rt do you want me to make internal release of this branch for testing or are you fine with emulators?

**Comment by s77rt** [comment]
@mountiny I'm fine with emulators. Unfortunately we will have to discuss some more before moving with this PR
1. How much performance are we actually gaining by using Freeze, is it that significant?
2. We have a couple of issues that are on hold in favor of the removal of Freeze, how are we supposed to handle those?

Besides the custom flat list here is a workaround, we have a bug report for that https://github.com/Expensify/App/issues/14125

**Comment by mountiny** [comment]
Essenatilly the implementation of the refactor means we will be pushing many reports to the stack, they wont be reused so if you will use app for a long time you will have a lot of pages. I think we need to use the Freeze for ReportScreens at least for now to ensure that those are not being rerendered in the background.

I am sure also we can work with SWM to fix the react-freeze issues, they are maintainers of the library

**Comment by s77rt** [comment]
@mountiny 

- On Android - Chrome it's causing the issue pointed by @parasharrajat on https://github.com/Expensify/App/pull/16581#issuecomment-1516470982. We are fixing this by not using Freeze.
- On Android - Native it's causing https://github.com/Expensify/App/issues/14125. We are fixing this by applying a workaround.

**Comment by mountiny** [comment]
Thanks, I understand that.

I think what we need to do is completely test what will be influenced (performance mainly) if we remove the freeze completely. Especially Android app hopes to get better performance with dropping the use of drawer so we need to be cautious with what will be the performance consequence of these changes. 

Could anyone try to some quick performance audit for removing the freeze? at least in terms of FPS?

**Comment by mountiny** [comment]
Asked in Slack to discuss this https://expensify.slack.com/archives/C01GTK53T8Q/p1682548765642099

**Comment by staszekscp** [comment]
I think in the `SidebarLinks` we would not need the `Freeze` component so much, because we will not push more than one `SidebarLinks` screen on the stack. There shouldn't be a huge amount of updates there either. 

On the other hand I think it is worth to keep 'ReportScreen` wrapped in the `Freeze` component, because there may be multiple chats opened, and their re-renders could cause performance harm, especially on slower devices.

An alternative solution that would work to fix this specific Android web issue is to leave the Freeze component, but add a check to see if the screen transition finished. I haven't tested that, but if you prefer this kind of solution I may check this out.

**Comment by mountiny** [comment]
So I feel like:
- if we are going to keep the Freeze for the ReportScreen then there is no benefit in removing it from the Sidebar too

We can work with SWM to fix the react-freeze issues, they are keen to fix anything which is wrong with it

**Review by mountiny — COMMENTED** [review]
@parasharrajat @s77rt @staszekscp @neil-marcellini after the rect discussion I think we should be bale to use freeze for the report screen but skip its usage for the sidebar, what do you think?

**Comment by s77rt** [comment]
@mountiny I'm okay with keeping Freeze I'm just questioning how to handle the issues that we assumed will be gone after navigation reboot. However I would still discourage the use of Custom FlatList, I would really prefer to keep a bug than apply a workaround.

**Comment by neil-marcellini** [comment]
So, is this ready for review again? Please message me on NewDot when it is.

**Review by mountiny — COMMENTED** [review]
@s77rt Can you please give this one a review again assuming we will keep the Freeze so we know clearly what we need to address, got a bit lost in the conversation cc @staszekscp thank!

File: src/components/FlatList/index.android.js
Comment (s77rt) [inline src/components/FlatList/index.android.js:1]: Let remove the CustomFlatList and just use the react-native FlatList

**Review by s77rt — CHANGES_REQUESTED** [review]
One additional change request. And please address the other unresolved comments.

**Comment by mountiny** [comment]
@staszekscp lets follow up on this one next week

**Comment by mountiny** [comment]
Closing this PR in favour of the final one we have here https://github.com/Expensify/App/pull/19263, the changes should be incorporated there.

**Pull request closed without merging**
